query-connector/src/app/backend/query-building.ts

```diff
@@ -28,3 +31,87 @@ export async function getSavedQueryDetails(queryId: string) {
     console.error("Error retrieving query", error);
   }
 }
+
+/**
+ * Backend handler function for upserting a query
+ * @param queryInput - frontend input for a query
+ * @param queryName - name of query
+ * @param author - author
+ * @param queryId - a queryId if previously defined
+ * @returns - all columns of the newly added row in the query table
+ */
+export async function saveCustomQuery(
+  queryInput: NestedQuery,
+  queryName: string,
+  author: string,
+  queryId?: string,
+) {
+  const queryString = `
+    INSERT INTO query
+      values($1, $2, $3, $4, $5, $6, $7, $8, $9)
+      ON CONFLICT(id)
+      DO UPDATE SET
+        query_name = EXCLUDED.query_name,
+        conditions_list = EXCLUDED.conditions_list,
+        query_data = EXCLUDED.query_data,
+        author = EXCLUDED.author,
+        date_last_modified = EXCLUDED.date_last_modified
+      RETURNING id, query_name;
+    `;
+  const { queryDataInsert, conditionInsert } =
+    formatQueryDataForDatabase(queryInput);
+
+  const NOW = new Date().toISOString();
+  try {
+    const dataToWrite = [
+      queryId ? queryId : randomUUID(),
+      queryName,
+      queryDataInsert,
+      conditionInsert,
+      author,
+      NOW,
+      NOW,
+      DEFAULT_TIME_WINDOW.timeWindowNumber,
+      DEFAULT_TIME_WINDOW.timeWindowUnit,
+    ];
+    const result = await dbClient.query(queryString, dataToWrite);
+    if (result.rows.length > 0) {
+      return result.rows as unknown as QueryDetailsResult[];
+    }
+    console.error("Query save failed:", dataToWrite);
+    return [];
+  } catch (error) {
+    console.error("Error saving new query", error);
+  }
+}
+
+function formatQueryDataForDatabase(frontendInput: NestedQuery) {
```

Comment:
Decided to store the data in the `query_data` column as `conditionId > valueSetId > ValueSet`. Think this gives us the most flexibility for further manipulation in the backend, but can change this as needed